Add in-place and genesis migrations (#205)
cosmos/ibc-go: docs/migrations/ibc-migration-043.md

NOTE: Updating to the `v0.43.0` SDK release and then running `go mod tidy` will cause a downgrade to `v0.42.0` in order to support the old IBC import paths.
Update the import paths before running `go mod tidy`.  

## Chain Upgrades

Chains may choose to upgrade via an upgrade proposal or genesis upgrades. Both in-place store migrations and genesis migrations are supported. 

> **WARNING**: Please read at least the quick guide for [IBC client upgrades](../ibc/upgrades/README.md) before upgrading your chain. It is highly recommended to not change the chain id during an upgrade, otherwise you must follow the IBC client upgrade instructions.

Author: ```suggestion
**WARNING**: Please read at least the quick guide for [IBC client upgrades](../ibc/upgrades/README.md) before upgrading your chain. It is highly recommended you do not change the chain-ID during an upgrade, otherwise you must follow the IBC client upgrade instructions.
```

Both in-place store migrations and genesis migrations will:

> - update the solo machine client state to using the v2 solo machine protobuf defintion

Author: ```suggestion
- migrate the solo machine client state from v1 to v2 protobuf definitions
```

- prune all solo machine consensus states
- prune all expired tendermint consensus states

### In-Place Store Migrations

In place store migrations will automatically be run after the binary is swapped and the application is started again during an upgrade proposal. 

### Genesis Migrations

To perform genesis migrations, the following code must be added to your existing migration code.

```go
// add imports as necessary
import (

>     "github.com/cosmos/cosmos-sdk/codec"

Author: ```suggestion
```

    ibcv100"github.com/cosmos/ibc-go/modules/core/legacy/v100"
    ibchost "github.com/cosmos/ibc-go/modules/core/24-host"
)

...

// add in migrate cmd function
newGenState, err = ibcv100.MigrateGenesis(clientCtx.InterfaceRegistry, newGenState, genDoc.GenesisTime)
if err != nil {
    return err
}
```

> 

Author: untested, will test when I do upgrade testing and update accordingly

### In-Place Store Migrations

The new chain binary will need to run migrations in the upgrade handler. The fromVM (previous module version) for the IBC module should be 1. This will allow migrations to be run for IBC updating the version from 1 to 2.

Ex:
```go
app.UpgradeKeeper.SetUpgradeHandler("my-upgrade-proposal",
        func(ctx sdk.Context, _ upgradetypes.Plan, _ module.VersionMap) (module.VersionMap, error) {
            // set max expected block time parameter. Replace the default with your expected value
            // https://github.com/cosmos/ibc-go/blob/release/v1.0.x/docs/ibc/proto-docs.md#params-2
            app.IBCKeeper.ConnectionKeeper.SetParams(ctx, ibcconnectiontypes.DefaultParams())

            fromVM := map[string]uint64{
                ... // other modules
                "ibc":          1,
                ... 
            }   
            return app.mm.RunMigrations(ctx, app.configurator, fromVM)

>         })      

Author: tested

```

### Genesis Migrations

To perform genesis migrations, the following code must be added to your existing migration code.

```go
// add imports as necessary
import (
    ibcv100 "github.com/cosmos/ibc-go/modules/core/legacy/v100"
    ibchost "github.com/cosmos/ibc-go/modules/core/24-host"
)

...

// add in migrate cmd function
// expectedTimePerBlock is a new connection parameter
// https://github.com/cosmos/ibc-go/blob/release/v1.0.x/docs/ibc/proto-docs.md#params-2
newGenState, err = ibcv100.MigrateGenesis(newGenState, clientCtx, *genDoc, expectedTimePerBlock)
if err != nil {
    return err 

> }

Author: tested, got as far as I could. Final confirmation is blocked on upstream changes (SDK is moving protobuf definition for upgrade/gov from v1beta1 to v1 and the current pr has minor bugs)

I confirmed the migration updated the pr, as far as I can tell, everything works on the IBC end, although I didn't get a chance to test that IBC clients were successfully upgraded. At the very least the chain started fine
